### ISSUE TITLE
drive: Fix 403 rateLimitExceeded errors

### ISSUE THREAD

**Issue by ChrisSLT** [ncw/rclone#20]
This is a great product, thank you!  Wish I could contribute actual code, but for now can only contribute testing results.  It's great to be able to create simple scripts that interact with these different cloud storage services.  I'm surprised this seems to be one of the few cross platform command line tools doing this.

I'm using it mostly with Google Drive for now but plan to use with S3 as well.  With Google Drive, I'm experiencing the error I think you are aware of when it operates over several GB worth of data and a large directory tree - Drive seems to not like all of the requests or something and it throws an error.

Interestingly, when I map Google Drive using Expandrive, and then use rsync, I don't get this error.  This may be because Expandrive slows down the requests due to it's own internal workings?  There are other problems with mapping a drive and then using the command line so I'd much rather not have to involve the drive mapping and just use the command line.

If there is some way I can help please let me know.

**Comment by ncw** [comment]
Thanks for the feedback. You can certainly help with testing things out!

I'd like to fix the Google Drive problems as apart from those errors it works really well. The errors are caused by rclone doing too many requests per second as far as I can see.   Here is what my quota for rclone says, so that should allow 250 requests/second/user.

    FREE QUOTA  10,000,000 requests/day
    REMAINING  9,933,268 requests/day 99.33% of total
    PER USER LIMIT  250 requests/second/user

I can try raising the number (I've already upped it from 100/s) or I could put some deliberate delays into rclone.  However I don't really think rclone is doing 250 requests per second.  It might be doing 25 in 100mS.

I've just raised the quota to 1000 requests/s/user.  I don't know how long it takes to take effect, but it would be interesting if you have another go in a few hours.

Next thing to try would be to pace the rclone requests so they always come more than 10 mS apart (say) which would be quite easy but require a bit of extra code.

As a workaround you can just run the rclone command again (and again) until it exits with no error - it will get there in the end - that is what I do! (I use Google Drive quite a lot.)

Another thing you could try is setting `--checkers 1` and `--transfers 1` - that might help.  Would be an interesting data point.

**Comment by ChrisSLT** [comment]
Hi - happy to test and report back.

However, based on my experience using Expandrive and rsync with s3/Drive/Dropbox, etc - and it all working pretty well even over large directory trees, I like your idea and think it would be worthwhile as a strategy to have it send requests with a 10ms delay or something like that between each one.

**Comment by ChrisSLT** [comment]
I just got a successful ls processed with no errors by the way

**Comment by ncw** [comment]
Hmm, a quick experiment shows that even waiting 250ms between queries isn't enough to avoid the 403 rate limit exceeded error.

I think I'll have to implement the exponential backoff and retry algorigthm here

https://developers.google.com/drive/web/handle-errors

As the quota setting in the drive control panel obviously isn't doing anything!

**Comment by ChrisSLT** [comment]
OK - just so I'm clear - I WAS able to run a ls command on my entire google drive account and it succeeded without error AFTER you raised the quota to 1000 requests.  Just FYI... let me know what further testing I can do.

**Comment by ncw** [comment]
I spent some time making an algorithm to fix this and I hope I've got it right in v1.09.  Please can you have a go and let me know if it isn't working for you.

Thanks for the bug report

Nick

**Comment by ChrisSLT** [comment]
Hi Nick,
Thanks for your work on this!

 Before I give this a go I wanted to get the test parameters set -

-Is it your sense that with Google Drive now there can be 10K+ files and it
will sync w/o errors?

-What about with Amazon S3, have you got it returning more than 10K files
now?

-Lastly, have you implemented the max size or include/exclude stuff like
rsync that you want me to test?

thanks

PS - still enjoying Hugo - what do you think about it?

Chris


On Sat, Feb 7, 2015 at 6:00 PM, Nick Craig-Wood <notifications@github.com>
wrote:

> I spent some time making an algorithm to fix this and I hope I've got it
> right in v1.09. Please can you have a go and let me know if it isn't
> working for you.
>
> Thanks for the bug report
>
> Nick
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ncw/rclone/issues/20#issuecomment-73387614>.
>



-- 
347.708.6549

**Comment by ncw** [comment]
- Is it your sense that with Google Drive now there can be 10K+ files and it
will sync w/o errors?

Yes that should work fine now (cross fingers!)

- What about with Amazon S3, have you got it returning more than 10K files
now?

Not yet.  That is a separate issue #22 . I'm finding rather a lot of bugs in the S3 library at the moment...

- Lastly, have you implemented the max size or include/exclude stuff like rsync that you want me to test?

No, not yet. If you want you can make an issue with exactly which features you would like - that will put it firmly on my radar!

It is the google drive performance that should be really different in v1.09.

**Comment by ChrisSLT** [comment]
Ok thanks will do

Re bugs in the s3 library, I'm not surprised. I'm using the aws cli which I
think uses python boto and sync works but it seems like they still have a
bit of work to do. What library are you using?

**Comment by ncw** [comment]
https://github.com/awslabs/aws-sdk-go is what I'm using.  It was working quite well until recently but alas now seems very broken!

**Comment by ChrisSLT** [comment]
Ah
This is probably unlikely, but is there any way you could use boto given
its in python? It's mature and officially supported by amz

**Comment by ncw** [comment]
I have decided to fix this with my old s3 library which does work.  Expect an update shortly!

**Comment by ChrisSLT** [comment]
Oh man I need to get better at using Github.  

Anyway, I tested this with Google Drive using 1.09 per your request.  File structure only 2GB, but it worked great first time.

**Comment by ncw** [comment]
Super, thanks!

**Comment by ChrisSLT** [comment]
Uh - oh - tried running sync command from local file system to Google Drive and immediately got 403 errors.  Using rclone 1.10

**Comment by ChrisSLT** [comment]
FYI, it still apparently uploaded files.  It just reports "Received error: Upload failed - retry: googleapi: Error 403: User rate limit exceeded, userRateLimitExceeded - retrying 1/10" for every single file.

I then ran a ls command and that worked fine.

**Comment by ncw** [comment]
Yes that is what it does - hopefully it didn't report any errors in the stats printed at the end.

When I try it I see it upload 20-30 files very quickly then rate limit, so get quite a few retries then start uploading quickly again.  That seems to be google's rate limiting in action.  It doesn't seem possible to upload more than about 2 files a second.

**Comment by ChrisSLT** [comment]
Ah, got it. It did complete uploading the 8K small text files.

Seems to be a nasty problem w/ the Google API...  takes a good long while
doesn't it.


On Mon, Feb 16, 2015 at 5:06 AM, Nick Craig-Wood <notifications@github.com>
wrote:

> Yes that is what it does - hopefully it didn't report any errors in the
> stats printed at the end.
>
> When I try it I see it upload 20-30 files very quickly then rate limit, so
> get quite a few retries then start uploading quickly again. That seems to
> be google's rate limiting in action. It doesn't seem possible to upload
> more than about 2 files a second.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ncw/rclone/issues/20#issuecomment-74484919>.
>



-- 
347.708.6549

**Comment by klokan** [comment]
@ncw thanks for great rclone soft!

Just an idea: if Google Drive after fast uploading of 20-30 files triggers the rate limit to 1-2 files per second, what about to request a new accessToken and continue with fast uploading of another 20-30 files and try to repeat this? The upload of 50.000 of small files takes now hours and hours - and I wonder if this trick could help. What do you think?

**Comment by ncw** [comment]
@klokan Interesting idea, I suspect that it won't work though - even if you get a new access token, google still knows who you are because you have to use the refresh token they gave you to get a new access token. So I think there is some underlying identity that google will keep track of instead.

As for lots of small file uploading - yes this is annoying.  Unfortunately I think it is an intrinsic part of drive.

I just checked my quotas for rclone - google must have put them up automatically!  I can't believe that you are doing more than 1000 requests/second so I think the rate limiting must be part of drive.

```
Free quota	
1,000,000,000 requests/day
Per-user limit	
1,000 requests/second/user
```

**Comment by klokan** [comment]
Essential is to know on what level is the "upload max. 1-2 files / second" rate limit applied:
- IP address / established network connection / router on certain HTTP POST / PUT
- Access Token
- Refresh Token
- Google Account

To try it:
Would it make sense to start multiple `rclone` processes in parallel - to test the Access token limit?
To try Refresh tokens for single user - one could configure storage aliases in rclone for the same drive, right?
If it does not help then probably use multiple users (multiple Google Accounts) for the upload - this must run, unless the rate limit is on IP address network level, correct?
Otherwise different users on different computers.

If we know where the limit really is, it should be possible to avoid it with multiple Access tokens / Refresh tokens / Users / Computers for uploading very large number of files faster...

Maybe a command-line trick like starting rclone via xargs -n 25 on the list of files could make the workaround without a need for coding? Your feedback on these ideas would be great... Thanks!

**Comment by klokan** [comment]
> "There is a lower limit specifically for uploads, and it's per-user across all apps. Will see if we can post the limit in docs."
From [Steve Bazyl](http://stackoverflow.com/users/1090284/steve-bazyl) (Google Inc.)
http://stackoverflow.com/questions/17536416/raising-google-drive-api-per-user-limit-does-not-prevent-rate-limit-exceptions

**Comment by timwhite** [comment]
Seems I keep hitting this limit :(
And that stackoverflow never got clarification from Google as to what the limit is.
I'm currently using ```rclone sync --size-only --transfers=1 --bwlimit 300``` in an attempt to upload 900Gb to Google Drive worth of photos. Most of it's done, so it's just the syncing of changes and deletions.
Any suggests from others as to what settings they have for Google Drive that help reduce the rate limits errors? Most times retry 3 or 4 appears to work.

**Comment by zeshanb** [comment]
bwlimit option needs to have M for megabytes after number e.g. 300M. This
number is limiting consistent transfer speed. The advertised transferred
speeds on consumer connections degrade over the length of upload of single
item and at times fluctuate a lot.  These other options will help also.

–transfers=N
Another good thing is to limit number of transfers at the same time. The
default is 4. On consumer connections set this to 1.

–timeout=TIME
In long transfers the connection times out periodically. Either on client
or service side.

Maybe the sync in recent version of rclone works as sync in rsync. That's a
huge gotcha for most users.

**Comment by timwhite** [comment]
bwlimit default unit is kBytes/s, so for me it's 300kBytes/s which is good for my connection.
I'm already limiting to 1 transfer at a time, still getting API ratelimites.

I'm not sure that it's timing out, that wouldn't give me the ratelimiting issue.

Also, I'm using sync with --size-only to ensure the sync isn't having to compute checksums for every file. (Also, it can't work like rsync due to the API's, if the file differs, we upload the whole file again)

**Comment by ncw** [comment]
@timwhite The idea in this issue might fix your problem #485 - comments on that issue please!
